File: quest/Assets/Orion/Runtime/Flight/LegStyle.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace Orion.Flight
{
    /* How a route sits on the city, and how each way of travelling looks.
     *
     * Heights come from rays dropped onto photogrammetry, which is a model of whatever is on
     * top: a tree that overhangs the street, an awning, a bridge deck. One such hit among points
     * 30 m apart is a spike in the line. Real streets do not jump twenty metres in thirty, so the
     * heights are cleaned before anything uses them. */

    public readonly struct LegStyle
    {
        const float MaxSlope = .3f;          // metres of height per metre along the route: steeper than any street

        /// <summary>The ribbon laid along the street: its width on the ground in metres, and the least it may look, as a
        /// fraction of how far off it is, so a route is still a route from two kilometres away. (The desktop gives the
        /// second in pixels of its window; these are the same angles.)</summary>
        public readonly float WidthM, MinWidth;
        /// <summary>Metres of mark and of gap along it, or 0 for an unbroken ribbon.</summary>
        public readonly float DashOn, DashOff;
        /// <summary>A wider, fainter one under it.</summary>
        public readonly bool Glow;
        public readonly float Opacity;
        /// <summary>How fast light moves along it in the direction of travel, metres a second.</summary>
        public readonly float FlowMps;
        /// <summary>Only a guess: drawn broken, faint, and as an arc over the city that does not pretend to be a street.</summary>
        public readonly bool Guess;

        const float PerPixel = .00104f;          // the angle of one pixel of the desktop's window

        LegStyle(float widthM, float minPx, float dashOn, float dashOff, bool glow, float opacity, float flowMps, bool guess = false)
        { WidthM = widthM; MinWidth = minPx * PerPixel; DashOn = dashOn; DashOff = dashOff; Glow = glow; Opacity = opacity; FlowMps = flowMps; Guess = guess; }

        /// <summary>Walking is footsteps, cycling long dashes, transit a ribbon and driving a wide road of light. A leg that
        /// is only an estimate is drawn broken and faint, so a guess never looks like a route.</summary>
        public static LegStyle For(string transport, bool estimated = false)
        {
            LegStyle s = transport switch
            {
                "cycle" => new LegStyle(5, 5.8f, 30, 8, false, .95f, 14),
                "transit" => new LegStyle(9, 8.5f, 0, 0, true, .95f, 24),
                "drive" => new LegStyle(11, 9.5f, 0, 0, true, .95f, 20),
                _ => new LegStyle(5, 5.5f, 8, 5, false, .95f, 9),          // walk; old plans predate the field
            };
            return estimated ? new LegStyle(s.WidthM, s.MinWidth / PerPixel, 14, 16, false, .45f, s.FlowMps, guess: true) : s;
        }

        /// <summary>The same points with their heights made believable: outliers replaced by the median of
        /// their neighbours, then the slope limited, then a light average. x and z are untouched.</summary>
        public static List<Vector3> SmoothHeights(List<Vector3> pts)
        {
            int n = pts.Count;
            if (n < 3) return new List<Vector3>(pts);
            var median = new float[n];
            var w = new List<float>(5);
            for (int i = 0; i < n; i++)
            {
                w.Clear();
                for (int k = Math.Max(0, i - 2); k < Math.Min(n, i + 3); k++) w.Add(pts[k].y);
                w.Sort();
                median[i] = w[w.Count >> 1];
            }
            void Run(int from, int to, int step)
            {
                for (int i = from; i != to; i += step)
                {
                    float d = new Vector2(pts[i].x - pts[i - step].x, pts[i].z - pts[i - step].z).magnitude * MaxSlope;
                    median[i] = Mathf.Min(median[i - step] + d, Mathf.Max(median[i - step] - d, median[i]));
                }
            }
            Run(1, n, 1); Run(n - 2, -1, -1);
            var o = new List<Vector3>(n);
            for (int i = 0; i < n; i++) o.Add(new Vector3(pts[i].x, (median[Math.Max(0, i - 1)] + median[i] * 2 + median[Math.Min(n - 1, i + 1)]) / 4, pts[i].z));
            return o;
        }
    }
}

File: src/fly/legStyle.ts
import { Vector3 } from 'three'
import type { Transport } from '../types'

/* How a route sits on the city, and how each way of travelling looks.
 *
 * Heights come from rays dropped onto photogrammetry, which is a model of
 * whatever is on top: a tree that overhangs the street, an awning, a bridge deck.
 * One such hit among points 30 m apart is a spike in the line. Real streets do not
 * jump twenty metres in thirty, so the heights are cleaned before anything uses
 * them: the same points feed the map lines, the flight camera and the headset's flight. */

const MAX_SLOPE = .3          // metres of height per metre along the route: steeper than any street

/** The same points with their heights made believable: outliers replaced by the median of
    their neighbours, then the slope limited, then a light average. x and z are untouched. */
export function smoothHeights(pts: Vector3[]): Vector3[] {
  const n = pts.length
  if (n < 3) return pts.map(p => p.clone())
  const y = pts.map(p => p.y)
  const median = y.map((_, i) => {
    const w = y.slice(Math.max(0, i - 2), Math.min(n, i + 3)).sort((a, b) => a - b)
    return w[w.length >> 1]
  })
  const run = (from: number, to: number, step: number) => {
    for (let i = from; i !== to; i += step) {
      const d = Math.hypot(pts[i].x - pts[i - step].x, pts[i].z - pts[i - step].z) * MAX_SLOPE
      median[i] = Math.min(median[i - step] + d, Math.max(median[i - step] - d, median[i]))
    }
  }
  run(1, n, 1); run(n - 2, -1, -1)
  return pts.map((p, i) => new Vector3(p.x, (median[Math.max(0, i - 1)] + median[i] * 2 + median[Math.min(n - 1, i + 1)]) / 4, p.z))
}

export type LegStyle = {
  /** The ribbon laid along the street: its width on the ground in metres, and the least it may
      look on screen in pixels, so a route is still a route from two kilometres up. */
  widthM: number
  minPx: number
  /** World metres of mark and gap along it, or null for an unbroken ribbon. */
  dash: { on: number; off: number } | null
  /** A wider, fainter one under it. */
  glow: boolean
  opacity: number
  /** How fast light moves along it in the direction of travel, and the comet with it, metres a second. */
  flowMps: number
  /** For 3D tubes: radius as a fraction of the drawing unit, and bead spacing if drawn as beads. */
  tube: number
  beads: number | null
}

/** Walking is footsteps, cycling long dashes, transit a ribbon and driving a wide road of light.
    A leg that is only an estimate is drawn broken and faint, so a guess never looks like a route. */
export function legStyle(transport: Transport, estimated = false): LegStyle {
  const base: Record<Transport, LegStyle> = {
    walk: { widthM: 5, minPx: 5.5, dash: { on: 8, off: 5 }, glow: false, opacity: .95, flowMps: 18, tube: .0022, beads: .0075 },
    cycle: { widthM: 5, minPx: 5.8, dash: { on: 30, off: 8 }, glow: false, opacity: .95, flowMps: 45, tube: .0026, beads: .016 },
    transit: { widthM: 9, minPx: 8.5, dash: null, glow: true, opacity: .95, flowMps: 90, tube: .0034, beads: null },
    drive: { widthM: 11, minPx: 9.5, dash: null, glow: true, opacity: .95, flowMps: 70, tube: .0042, beads: null },
  }
  const s = base[transport] ?? base.walk      // old plans on disk predate the field
  return estimated ? { ...s, dash: { on: 14, off: 16 }, glow: false, opacity: .45, beads: s.beads ?? .012 } : s
}
